File: my-solutions/0572-subtree-of-another-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:

        res = False
        def isSubRoot(a, b):
            if a and b:
                return isSubRoot(a.left, b.left) and isSubRoot(a.right, b.right) and a.val == b.val
            elif not a and not b:
                return True
            else:
                return False

        def dfs(root):
            nonlocal res
            if root:
                res = res or isSubRoot(root, subRoot)
                dfs(root.left)
                dfs(root.right)

        dfs(root)

        return res

File: my-solutions/0572-subtree-of-another-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool checkSubTree(TreeNode* root, TreeNode* subroot) {
        if(subroot && root) {
            // cout << root->val << " " << subroot->val << endl;
            return subroot->val == root->val && checkSubTree(root->left, subroot->left) && checkSubTree(root->right, subroot->right);
        } else if (!root && !subroot) {
            // cout << "null" << " " << "null" << endl;
            return true;
        } else {
            return false;
        }
    }
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(!root){
            return false;
        } else {
            if(checkSubTree(root, subRoot)) {
                // cout << root->val << " " << subRoot->val << endl;
                return true;
            } else {
                return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
            }
        }
    }
};
